### PR TITLE
feat(stages): checkpoint hashing stages into a new table

### PR DESCRIPTION
**Pull request by shekhirin**
Continuing the work on revamping the checkpoints. For hashing stages, we've been previously using separate DB tables to store the stage-specific checkpoints, now we can do it using just the `SyncStage` table.

For Merkle stage, see https://github.com/paradigmxyz/reth/pull/2701#discussion_r1196933358. After Merkle checkpoint is moved to the new unified table, we are free to remove the `SyncStageProgress` table completely.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
supportive of this,

pending @rkrasiuk

File: crates/stages/src/stages/hashing_account.rs [hunk 1]
```diff
@@ -168,27 +130,37 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
         }
         let (from_block, to_block) = range.into_inner();
 
+        let stage_checkpoint =
+            input.checkpoint.and_then(|checkpoint| checkpoint.account_hashing_stage_checkpoint());
+
+        if stage_checkpoint.is_some() {
+            debug!(target: "sync::stages::account_hashing::exec", checkpoint = ?stage_checkpoint, "Continuing inner account hashing checkpoint");
```
Comment (rkrasiuk) [inline crates/stages/src/stages/hashing_account.rs:137]: this debug log is valid only if the checkpoint range didn't change. pls move it below where we check that logic

File: crates/stages/src/stages/hashing_storage.rs [hunk 2]
```diff
@@ -89,27 +51,40 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
         }
         let (from_block, to_block) = range.into_inner();
 
+        let stage_checkpoint =
+            input.checkpoint.and_then(|checkpoint| checkpoint.storage_hashing_stage_checkpoint());
+
+        if stage_checkpoint.is_some() {
+            debug!(target: "sync::stages::storage_hashing::exec", checkpoint = ?stage_checkpoint, "Continuing inner storage hashing checkpoint");
```
Comment (rkrasiuk) [inline crates/stages/src/stages/hashing_storage.rs:58]: same as the previous comment

**Review by rkrasiuk — COMMENTED** [review]
did you try syncing this on sepolia/mainnet? ideal check would be to kill the app while it's in one of the hashing stages and restart to make sure it restores correctly

**Review by rkrasiuk — COMMENTED** [review]
did you try syncing this on sepolia/mainnet? ideal check would be to kill the app while it's in one of the hashing stages and restart to make sure it restores correctly

**Comment by shekhirin** [comment]
@rkrasiuk yep. Added 100ms sleep for hashing stages for ease of stop, synced to 100k mainnet blocks, dropped hashing stages, started node, stopped node while it executed account or storage hashing stage, then restarted.

File: crates/stages/src/stages/hashing_account.rs [hunk 3]
```diff
@@ -173,22 +135,33 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
...
+                .and_then(|checkpoint| checkpoint.account_hashing_stage_checkpoint());
+
+            let start_address = match stage_checkpoint {
+                Some(AccountHashingCheckpoint { address: address @ Some(_), from, to })
+                    // Checkpoint is only valid if the range of transitions didn't change.
+                    // An already hashed account may have been changed with the new range,
+                    // and therefore should be hashed again. 
+                    if from == from_block && to == to_block =>
+                {
+                    if stage_checkpoint.is_some() {
```
Comment (rkrasiuk) [inline crates/stages/src/stages/hashing_account.rs:149]: this is redundant now

File: crates/stages/src/stages/hashing_storage.rs [hunk 4]
```diff
@@ -94,22 +56,36 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
...
+                         storage,
+                         from,
+                         to ,
+                 })
+                // Checkpoint is only valid if the range of transitions didn't change.
+                // An already hashed storage may have been changed with the new range,
+                // and therefore should be hashed again. 
+                if from == from_block && to == to_block =>
+                    {
+                        if stage_checkpoint.is_some() {
```
Comment (rkrasiuk) [inline crates/stages/src/stages/hashing_storage.rs:75]: same here

File: crates/stages/src/stages/hashing_account.rs (hunk 3 above)
Comment (shekhirin) [reply crates/stages/src/stages/hashing_account.rs]: oh right thanks, removed

**Review by rkrasiuk — APPROVED** [review]
lgtm!

**Pull request merged**